feat: Add User Manager

jeevithakannan2 opened the pull request:
# Pull Request

## User manager to create / modify users 

## Type of Change
- [x] New feature
- [x] Documentation Update

## Description
Add a user manager script that can do the following
- Create a new user
- Delete a user
- Modify a user
  - Change password
  - Add user to group
  - Remove user from group
 
 - Also added docs for utilities 

## Testing
- Tested by adding removing modifying users.
- If root user is modified only the password change is reflected in the system adding to group or removing from group shows successful message but it did not have any affect in the system.
- Added a check if the user is trying to delete or modify a system user [ i.e Users created by apps, For example git has a user ]

## Checklist
- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no errors/warnings/merge conflicts.

nnyyxxxx commented on tabs/utils/user-manager.sh:
```diff
@@ -0,0 +1,263 @@
+            2) deleteUser ;;
+            3) modifyUser ;;
+            0) exit 0 ;;
+            *) printf "%b\n" "${RED}Invalid option. Please [Enter] to try again.${RC}"; read -r ;;
```
```bash
            *) printf "%b\n" "${RED}Invalid option. Please [Enter] to try again.${RC}"; read -r ;;
```
This is not posix compliant.

nnyyxxxx commented on tabs/utils/user-manager.sh:
```diff
@@ -0,0 +1,263 @@
+
+    promptPassword
+
+    $ESCALATION_TOOL useradd -m "$username" -g users -G wheel,audio,video -s /bin/bash
```
Use a safer method please.

nnyyxxxx commented on tabs/utils/user-manager.sh:
```diff
@@ -0,0 +1,263 @@
+        return
+    fi
+
+    $ESCALATION_TOOL userdel --remove "$username" 2>/dev/null
```
Use a safer method please.

nnyyxxxx replied:
For all user management commands they need to be wrapped in an ``if`` block with safer conditionals. @jeevithakannan2

jeevithakannan2 replied:
Give me more information and a sample code

nnyyxxxx replied:
> Give me more information and a sample code

Why would I do that? I gave you enough information. Should be simple enough to do.

jeevithakannan2 replied:
I don't understand what you meant by a safer method !!!

jeevithakannan2 replied:
All the if conditions are already defined before the commands !!

nnyyxxxx replied:
@jeevithakannan2 ``$ESCALATION_TOOL userdel --remove "$username" 2>/dev/null`` Needs to be put in an ``if`` block with safer conditionals, please read what I said.

nnyyxxxx replied:
> @jeevithakannan2 $ESCALATION_TOOL userdel --remove "$username" 2>/dev/null Needs to be put in an ``if``block with safer conditionals, please read what I said.
Other user related tasks as well.

jeevithakannan2 replied:
The above if conditions check whether it is okay to run the command. If not it will return back without executing the command

guruswarupa commented:
rename things to User Account Manager.

guruswarupa commented:
@jeevithakannan2 could you move these scripts to utils? + rename it to user account manager. whats user manager?

jeevithakannan2 commented:
It must be under utility it is not an application or system setup. And it should've been user account manager not user manager

nnyyxxxx commented on tabs/utils/test.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
```
Do not change this file.

jeevithakannan2 replied:
I was testing some functions and accidentally left that file there

nnyyxxxx commented on tabs/utils/tab_data.toml:
```diff
@@ -65,3 +65,26 @@ script = "monitor-control/scale_monitor.sh"
 matches = true
+
+[[data]]
+name = "User Account Manager"
```
This implementation in tab data looks good, don't change this.

jeevithakannan2 replied:
Well @guruswarupa wanted it to be user account manager and not user manager

guruswarupa commented:
> It must be under utility it is not an application or system setup. And it should've been user account manager not user manager

my man am i chatting in spanish? that's exactly what i asked u to do.

nnyyxxxx commented on tabs/utils/user-account-manager/add_to_group.sh:
```diff
@@ -0,0 +1,33 @@
+username=$(promptUsername "" "non-root") || exit 1
+user_groups=$(groups "$username" | cut -d: -f2 | sort | tr '\n' ' ')
+
+clear
```
Stop using clear so much.

jeevithakannan2 replied:
If you run the script you will know why it was used there

nnyyxxxx replied:
> If you run the script you will know why it was used there

I'm talking about in general, clear should not be used that much in interactive scripts + how are you going to debug errors / bugs with the terminal screen clearing 24/7. lmfao

nnyyxxxx commented on tabs/utils/user-account-manager/add_user.sh:
```diff
@@ -0,0 +1,23 @@
+echo "$username:$password" | $ESCALATION_TOOL chpasswd
+
+printf "%b\n" "${GREEN}User $username created successfully${RC}"
+printf "%b\n" "${GREEN}To add additional groups use Add User To Groups${RC}"
```
Always call checkEnv at the end of the script, if you are also calling ESCALATION_TOOL then add checkEscalationTOol under checkEnv aswell.

jeevithakannan2 replied:
It is called in . ./utility_functions.sh

nnyyxxxx replied:
@jeevithakannan2 That does not matter. You should always be calling checkEnv at the end of scripts.

jeevithakannan2 replied:
It's overkill

nnyyxxxx replied:
> It's overkill

Not exactly, If only checkEscalationTool is being called and any other function than those functions along with checkEnv NEED to be called at the bottom and checkEnv NEEDS to be always first at the bottom, you can use ``clear`` to clear the terminal output.

nnyyxxxx replied:
> > It's overkill
> 
> Not exactly, If only checkEscalationTool is being called and any other function than those functions along with checkEnv NEED to be called at the bottom and checkEnv NEEDS to be always first at the bottom, you can use `clear` to clear the terminal output.

@jeevithakannan2

nnyyxxxx commented on tabs/utils/user-account-manager/utility_functions.sh:
```diff
@@ -0,0 +1,99 @@
+    done
+}
+
+checkEscalationTool
```
As I said above checkEnv always has to be called first.

jeevithakannan2 replied:
checkCommandRequirements 'curl groups sudo'
  checkPackageManager 'apt-get nala dnf pacman zypper yum xbps-install nix-env'
  checkCurrentDirectoryWritable
  checkSuperUser
  checkDistro
  checkAURHelper
  
  No need of these extra functions to be called

nnyyxxxx replied:
Again, checkEnv needs to be called for checkEscalationTool and other functions to work properly please stop trying to debate me on this, thanks. @jeevithakannan2

guruswarupa commented:
@jeevithakannan2 rename user-account-manager to User Account Manager. that's being displayed to end users. why casing it all to lower?

nnyyxxxx commented:
@jeevithakannan2 Please unresolve those reviews, they are not resolved in the slightest.

jeevithakannan2 commented:
> @jeevithakannan2 Please unresolve those reviews, they are not resolved in the slightest.


It's resolved

nnyyxxxx commented:
@jeevithakannan2 Check the comment I added in the review. I explained it more.

jeevithakannan2 commented:
> @jeevithakannan2 Check the comment I added in the review. I explained it more.

Added them like you wanted. not going to ague over this.  Wasted my time on this crap

jeevithakannan2 commented:
Don't ask to me squash the commits. I'm not going to do that !! Not going to waste my time on this again

guruswarupa commented:
@jeevithakannan2 we reviewers ask you to do certain things so that the script follows the uniformity in scripts as well as is free from bugs which could happen by not calling those functions. now squashing your commits is necessary because chris wont accept it when u have so many commits on such a small file. if you dk how to, we'll teach you. dont worry.

guruswarupa commented:
@jeevithakannan2 open your fork in vs code. go to the respective branch where this is present. then, in terminal type ```git rebase -i HEAD~8  ``` then using editor keep first one as pick and change rest from pick to s. now another window will open. here remove all commit messages other than the first one. and again save and exit. and then type ```git push origin branchname --force```
